### PR TITLE
Fix k-Caching in presence of iteration spaces

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

Glocal Iteration spaces were introduced. This has consequences w.r.t to caching. If a stage with an iteration space caches a field, subsequent reads in another stage may "miss" the cache if this other stage has no iteration space or an iteration space outside the first stages'. In these cases, caching needs to be disabled. 

### Testing

There are various simple tests with two stages introduced which check the correct behavior of the caching in presence of iteration spaces.

### CODE REVIEW

File: dawn/src/dawn/IIR/Stage.cpp [hunk 1]
```diff
@@ -347,5 +347,26 @@ void Stage::setIterationSpace(const IterationSpace& value) { iterationSpace_ = v
 
 const Stage::IterationSpace& Stage::getIterationSpace() const { return iterationSpace_; }
 
+bool Stage::hasIterationSpace() const {
+  return iterationSpace_[0].has_value() || iterationSpace_[1].has_value();
```
Comment (twicki) [inline dawn/src/dawn/IIR/Stage.cpp]:
```suggestion
return std::any_of(iterationSpace_.cbegin(),iterationSpace_.cend(), [](const auto& p) { return p.has_value(); })
```

File: dawn/src/dawn/Optimizer/PassSetCaches.cpp [hunk 2]
```diff
@@ -227,6 +227,7 @@ bool PassSetCaches::run(const std::shared_ptr<iir::StencilInstantiation>& instan
         const auto& fields = ms.getFields();
         for(const auto& AccessIDFieldPair : fields) {
           const iir::Field& field = AccessIDFieldPair.second;
+          // printf("%s: %d", ms.getMetadata().getNameFromAccessID(field.getAccessID()).c_str(), field.getIntend());
```
Comment (twicki) [inline dawn/src/dawn/Optimizer/PassSetCaches.cpp]: rm

File: dawn/src/dawn/Unittest/IIRBuilder.cpp [hunk 3]
```diff
@@ -183,6 +183,15 @@ IIRBuilder::Field CartesianIIRBuilder::field(std::string const& name, FieldType
   return {id, name};
 }
 
+IIRBuilder::Field CartesianIIRBuilder::tmpField(std::string const& name, FieldType ft) {
```
Comment (twicki) [inline dawn/src/dawn/Unittest/IIRBuilder.cpp]:
```suggestion
IIRBuilder::Field CartesianIIRBuilder::tmpField(const std::string& name, FieldType ft) {
```

File: dawn/test/unit-test/dawn/Optimizer/Passes/TestPassCaching.cpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+              1); // we expect the instantiation to contain one stencil...
+  ASSERT_TRUE(stencil->getStencils().at(0)->getChildren().size() == 1); //... with one multistage
+
+  // check if the cache was set correctly (we expect a k cache fill on tmp) 
+  const auto& multiStage = *stencil->getStencils().at(0)->getChildren().begin();
+  ASSERT_TRUE(multiStage->isCached(tmp.id));
+  ASSERT_TRUE(multiStage->getCache(tmp.id).getIOPolicy() == iir::Cache::IOPolicy::fill);
+  ASSERT_TRUE(multiStage->getCache(tmp.id).getType() == iir::Cache::CacheType::K);
+
+  //   std::ofstream of("check_kcache.cpp");
```
Comment (twicki) [inline dawn/test/unit-test/dawn/Optimizer/Passes/TestPassCaching.cpp]: rm

File: dawn/src/dawn/Optimizer/PassSetCaches.cpp [hunk 5]
```diff
@@ -248,6 +249,32 @@ bool PassSetCaches::run(const std::shared_ptr<iir::StencilInstantiation>& instan
...
+              break;
+            }
+            stageIter++;
+          }
+
+          //check if subsequent uses are compatible
+          auto firstUsageWithGlobalIndex = ms.getChildren().begin();
+          std::advance(firstUsageWithGlobalIndex, stageIter);
+          bool incompatibleIterationSpaces = false;
+          for (auto it = firstUsageWithGlobalIndex; it != ms.getChildren().end(); it++) {
```
Comment (twicki) [inline dawn/src/dawn/Optimizer/PassSetCaches.cpp]: is this loop over all other stage? It seems like the check that the field in question is used is missing?

Comment (mroethlin) [reply dawn/src/dawn/Optimizer/PassSetCaches.cpp]: loop is over all subsequent stages. that check that field is in use was indeed missing, good catch!

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**